### PR TITLE
API Design

### PR DESCRIPTION
**Pull request by lukemelia**
[Rendered](https://github.com/adopted-ember-addons/ember-keyboard/blob/rfc/api-change/API-DESIGN.md)

We're implementing in this PR as well. Here's the check list:

- [x] on-key helper basic implementation
- [x] on-key modifier support for input, textarea, select elements
- [x] on-key modifier defaults to triggering click if handler function is not specified
- [x] deprecate `on-keyboard` modifier
- [x] deprecate `keyboard-shortcut` modifier
- [x] deprecate `keyboard-press` component
- [x] decide on new responder API (currently `has`/`trigger` via Evented mixin) that leverages `isKey` function
- [x] update API-DESIGN.md with new responder API https://github.com/adopted-ember-addons/ember-keyboard/blob/rfc/api-change/API-DESIGN.md#responder-api
- [x] implement basic `isKey` function https://github.com/adopted-ember-addons/ember-keyboard/pull/135/commits/d50035c1e2cec300755ccbeec05b49f2926f2e20
- [x] move new API to new responder API
- [x] preserve fallback to old responder API with deprecation warning
- [x] Support for `stopPropagation`/`stopImmediatePropagation` from `on-key` helper
- [x] Support for `stopPropagation`/`stopImmediatePropagation` from `on-key` modifier
- [x] implement `if-key` helper
- [x] implement `_all` support in `isKey` function
- [x] implement `key` mapping support in `isKey` function
- [x] assertion for callback argument to `on-key` helper being a function
- [x] implement decorator support for ES6 classes
- [x] implement decorator support for classic classes
- [x] document decorator usage
- [x] deprecate mixins
- [x] documentation review
- [x] write a V5 -> v6 migration guide
- [ ] ship v6.0.0-beta.1
- [ ] ship v6.0.0

### CODE REVIEW

**Comment by mattmcmanus** [comment]
I like where this is headed. I'd be in favor of proposal 3 for a couple reasons. 

1. The `key` mode will be the intuitive default and from my uneducated hunch, will be used by the majority of users
3. I like defaulting to `keydown`. 
2. From an end-user perspective, I like using the shorter and potentially more expressive "Alt+c" approach. As for key/code normalization and other things, have you considered backing that by a  library from the larger JS ecosystem like [Mousetrap](https://craig.is/killing/mice)? It seems like it's worked a lot of that stuff out and has some extra feature which would be nice to have.

There are some things that feel intuitive to me about these changes though.

* Using `on` as the attribute when changing to keyup doesn't feel ideal. `on` is starting to feel a bit overloaded. What about `when="keyup"`?
* Can you flip the order of the positional params? Having the function first, then the key combo feels backwards to me. 
* I don't love the name `on-keyboard`. Why not use `on-keypress` or even `on-key`? 
* It's not immediately obvious to me what the `keyboard-shortcut` modifier does? It requires some other `{{on}}` action to be wired up to work? Will it validate that something exists?
* Having a component and a modifier with the same name feels risky. 
  * What about the component being named `keyboard-shortcut` since that what it's setting up?
  * Another thought, what about precedence when a modifier and component potentially conflict? I need to think about this some more, but it seems like having just the modifier or just the component would simplify the mental model. Mousetraps documentation provides some ideas for this: https://craig.is/killing/mice#api.handleKey

Ok. That's all I've got!

**Comment by lukemelia** [comment]
@mattmcmanus Thanks for the excellent feedback:

> As for key/code normalization and other things, have you considered backing that by a library from the larger JS ecosystem like Mousetrap? It seems like it's worked a lot of that stuff out and has some extra feature which would be nice to have.

I looked into mousetrap and keymaster to see how they handled key vs code and whether we could/should use them. In both cases, I found that they rely on now-deprecated keyboard event properties (`which` and `keyCode`). e.g. https://github.com/ccampbell/mousetrap/issues/474. For that reason, I don't think that's a good path at the moment.

> Using `on` as the attribute when changing to keyup doesn't feel ideal. on is starting to feel a bit overloaded. What about `when="keyup"`?

Good point, or maybe `event="keyup"`.

> Can you flip the order of the positional params? Having the function first, then the key combo feels backwards to me.

If we do this, we lose the ability to attach additional shortcuts to one action by adding key combo strings as additional arguments. I think that flexibility is worth it, but absent that I agree the order would read better flipped.

> I don't love the name `on-keyboard`. Why not use `on-keypress` or even `on-key`?

I think `on-keypress` is out because `keypress` is the name of a deprecated event, so there is potential for confusion. `on-key` could work. Personally I like `on-keyboard` slightly more because the class of events we're handling are Keyboard events in W3C parlance.

> It's not immediately obvious to me what the `keyboard-shortcut` modifier does? It requires some other `{{on}}` action to be wired up to work? Will it validate that something exists?

It triggers a click on the element it is attached to. I'll update the doc to clarify this.

> Having a component and a modifier with the same name feels risky. What about the component being named `keyboard-shortcut` since that what it's setting up?

I can see what it would be confusing. The difference between the two is that the component listens and is triggerable whenever it is rendered while the modifier only triggers when the element is focused. I think we need both, though they don't necessarily need to be named the same thing.

> Another thought, what about precedence when a modifier and component potentially conflict? I need to think about this some more, but it seems like having just the modifier or just the component would simplify the mental model. Mousetraps documentation provides some ideas for this: https://craig.is/killing/mice#api.handleKey

It's not in this API document currently, but ember-keyboard has a bunch of existing support for managing priority and propagation that I think we would want to preserve as-is.

**Comment by Alonski** [comment]
@lukemelia Have you publically shared this on Discord and Twitter? Maybe others can chime in :)
I don't have much usage of this addon so not sure how much my thoughts will be good here :)

**Comment by lukemelia** [comment]
@Alonski Yes, I have. If you can help amplify, that would be great: https://twitter.com/lukemelia/status/1256439569181900800

**Comment by seanCodes** [comment]
Just to chime in, I'm also a fan of proposal 3.

I wonder though if `mode="code"` is even needed, since it seems the combo string itself can define the mode based on whether a key [code value](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code/code_values) is used. For example, `"f"` could indicate key mode and `alt+KeyC` could indicate code mode (which sounds backwards but follows the native [`KeyboardEvent` key code naming](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code/code_values)).

>> I don't love the name `on-keyboard`. Why not use `on-keypress` or even `on-key`?
>
> I think `on-keypress` is out because `keypress` is the name of a deprecated event, so there is potential for confusion. `on-key` could work. Personally I like `on-keyboard` slightly more because the class of events we're handling are Keyboard events in W3C parlance.

+1 to `{{on-key}}` vs `{{on-keyboard}}`. I had the same thought while reading through the RFC as `on-keyboard` was a bit verbose and clunky to read in the examples.

I’m a big fan of aligning with native terminology whenever possible, but in this case something like `on-key` or `on-keys` just seems so much more ergonomic and intuitive. 

>> Having a component and a modifier with the same name feels risky. What about the component being named `keyboard-shortcut` since that what it's setting up?
>
> I can see what it would be confusing. The difference between the two is that the component listens and is triggerable whenever it is rendered while the modifier only triggers when the element is focused. I think we need both, though they don't necessarily need to be named the same thing.

I didn't find `{{keyboard-shortcut}}` too confusing, but I think there might be a better name there as well. Maybe `{{keybinding}}` or `{{bind-to-key}}`? Or something more literal like `{{trigger-on-key}}`?

>>Can you flip the order of the positional params? Having the function first, then the key combo feels backwards to me.
>
>If we do this, we lose the ability to attach additional shortcuts to one action by adding key combo strings as additional arguments. I think that flexibility is worth it, but absent that I agree the order would read better flipped.

I also think flipped order would be ideal... Not only does it feel more natural, it’s also a pattern users are familiar with because the `{{on}}` modifier works the same way. Could multiple key combos not be comma or space separated instead? Or optionally an array? The `on` modifier requires multiple `{{on ...}}` statements for each event, so that would also be an option.

> > As for key/code normalization and other things, have you considered backing that by a library from the larger JS ecosystem like Mousetrap? It seems like it's worked a lot of that stuff out and has some extra feature which would be nice to have.
> 
> I looked into mousetrap and keymaster to see how they handled key vs code and whether we could/should use them. In both cases, I found that they rely on now-deprecated keyboard event properties (`which` and `keyCode`). e.g. [ccampbell/mousetrap#474](https://github.com/ccampbell/mousetrap/issues/474). For that reason, I don't think that's a good path at the moment.

It’s a bummer that there’s no other solid JS keyboard libraries out there that could be leveraged. Mousetrap has always been my favorite but it’s definitely outdated and neglected. [Hotkeys](https://github.com/jaywcjlove/hotkeys) (a continuation of Keymaster) seems to be the only one actively maintained, but it still uses `which`. I’m amazed that this is the only project trying to use the new API. (Maybe we create a vanilla library based on this code that could then be leveraged by this addon? 😄)

**Comment by lukemelia** [comment]
@seanCodes That's some excellent feedback!

> I wonder though if `mode="code"` is even needed, since it seems the combo string itself can define the mode based on whether a key code value is used. For example, "`f`" could indicate `key` mode and `alt+KeyC` could indicate `code` mode (which sounds backwards but follows the native KeyboardEvent key code naming).

I really like this idea. I just checked the [list of `code` values](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code/code_values) and it doesn't seem like there is any overlap. In addition, it means that people new to keyboard events can largely get what they need done without having to go down this particular rabbit hole right away and likely become confused and frustrated.

I am going to update the document to take this approach.

>I’m a big fan of aligning with native terminology whenever possible, but in this case something like `on-key` or `on-keys` just seems so much more ergonomic and intuitive.

OK, I find your and @mattmcmanus' feedback persuasive on this. I'll update.

>I also think flipped order would be ideal... Not only does it feel more natural, it’s also a pattern users are familiar with because the `{{on}}` modifier works the same way. Could multiple key combos not be comma or space separated instead? Or optionally an array? The on modifier requires multiple {{on ...}} statements for each event, so that would also be an option.

This is a good point. I will update to align to the style of the `{{on}}` modifier and we can defer binding multiple keys in one invocation until we see real-world needs.

> It’s a bummer that there’s no other solid JS keyboard libraries out there that could be leveraged. Mousetrap has always been my favorite but it’s definitely outdated and neglected. Hotkeys (a continuation of Keymaster) seems to be the only one actively maintained, but it still uses which. I’m amazed that this is the only project trying to use the new API.

Seriously. I think it's probably due to IE 11 not supporting these APIs, but still -- it's surprising.

>(Maybe we create a vanilla library based on this code that could then be leveraged by this addon? 😄)

Please do implement. ;-)

**Comment by lukemelia** [comment]
I've incorporated feedback from @mattmcmanus and @seanCodes into (Proposal 4)[https://github.com/adopted-ember-addons/ember-keyboard/blob/rfc/api-change/API-DESIGN.md#api-proposal-4-infer-key-vs-code-mode-default-to-listening-on-keydown-try-to-match-on--semantics-more-closely-use-on-key-for-shortcuts-too], which is my new favorite.

It has one not-yet-discussed idea in it as well, which is to use the same `{{on-key ...}}` modifier instead of `{{keyboard-shortcut ...}}` but infer the "trigger a click" behavior if the developer does not pass an action argument.

**Comment by mattmcmanus** [comment]
@lukemelia @seanCodes Love it! Proposal 4 feels so much more ergonomic and intuitive. 

Now my questions are: 
1. Will there be an easy way to query the list of currently available shortcuts? This would be used to support the `Shift+/` or `?` shortcut to bring up a list of all the currently available shortcuts
2. The simpler `alt+t` way of defining a shortcut opens up the ability to do sequences as well. I'm thinking along the lines of gmail's `g i` to go to the inbox
3. Is a codemod from the old API to the new even possible?

**Comment by bendemboski** [comment]
I very much like proposal 4, but have one suggestion that I think is relatively minor in terms of the work/implementation, but would make it quite a bit more flexible. I think that the core key-mapping engine that takes DOM events and recognizes key patterns should be exposed independently of the DOM event handling, since I think that's the core value that this would be providing. So in a template, it would allow something like

```hbs
{{!-- attach your own event handler using the {{on}} modifier --}}
<div {{on "keypress" (dispatch-key "alt+c" this.doThing)}}></div>

{{! use ember-on-helper }}
{{on-document "keydown" (dispatch-key "alt+x" this.doThing)}}

{{! use some third-party component API }}
<SomeComponent @onKey={{dispatch-key "alt+x" this.doThing}}/>
```

and in Javascript maybe

```javascript
function onEvent(e) {
  if (keyEventMatches(e, 'alt+x')) {
    this.handleAltX();
  }
}
```

Absolutely still provide the higher-level API you've specified as the primary API for the addon, but also expose these lower-level primitives so if you don't anticipate some particular pattern for capturing DOM events, it's not a blocker and user can still leverage the harder-to-duplicate DOM-event parsing logic.

**Comment by lukemelia** [comment]
@mattmcmanus wrote:
>Will there be an easy way to query the list of currently available shortcuts? This would be used to support the Shift+/ or ? shortcut to bring up a list of all the currently available shortcuts

Registered key handlers would be available via the the keyboard service, but I think it is likely that an interface like you've described would be maintained independently. Things like order, layout and a description of each key combos function would be impossible to standardize in a one-size-fits-most way.

@mattmcmanus wrote:
>The simpler alt+t way of defining a shortcut opens up the ability to do sequences as well. I'm thinking along the lines of gmail's g i to go to the inbox

True. This could be a cool future enhancement.

@mattmcmanus wrote:
>Is a codemod from the old API to the new even possible?

I don't think so. The issue that prompted this rethink was that we were conceptually mixing `key` and `code` treatment, so the existing behavior is something that we would not want to recreate. That said, since we are leaning toward brand new API, it could be possible to leave the existing API in place with the existing behavior and deprecate it. This would allow people to upgrade and incrementally move to the new API.

@bendemboski wrote:
>I think that the core key-mapping engine that takes DOM events and recognizes key patterns should be exposed independently of the DOM event handling, since I think that's the core value that this would be providing.

I like this idea a lot. It would also help us ensure this part of the codebase is well-isolated and can be covered by a thorough test suite. This low-level API would by necessity not participate in other ember-keyboard features like priority and propagation options but that doesn't obviate its utility. I think I would find `if-key` to be clearer than `dispatch-key` if I'm understanding the intended behavior clearly.

**Comment by bendemboski** [comment]
> I think I would find if-key to be clearer than dispatch-key if I'm understanding the intended behavior clearly.

Sure, makes sense. I called it `dispatch-key` because the helper took a whole table of key mappings, not just one. The only advantage I see to that is in reducing the number of DOM event listeners that are needed, but this seems like a pretty low priority given that I don't think people will be setting up, like, hundreds (or even tens) of key handlers. So if it's just defining one mapping, I agree that `if-key` seems clearer.

**Comment by optikalefx** [comment]
Chiming in here super late, sorry everyone. Had gotten pulled away from my upgrade which included this PR.

I'm a big fan of the evolution to Proposal 4. However I think I prefer the `mode='code'` approach from 3. 

I think it comes down to documentation at this point. "alt+c" and "alt+KeyC" aren't immediately obvious. Given just these you really don't know for sure what the difference between "alt+c" and "alt+KeyC" is. Users will need to go read the docs and discover what it means.

However, `"alt+c" mode="key"` and `"alt+c" mode="code"` does say exactly what is happening. But, requires you to know about key vs code in the first place. So perhaps either way you need to read more.

It's really down to "how much magic" do we want here `mode` is less magic. Is that better? Hard to say - I think the way of Octane has been to be less magic.

**Comment by lukemelia** [comment]
@optikalefx wrote:
>However, "alt+c" mode="key" and "alt+c" mode="code" does say exactly what is happening. But, requires you to know about key vs code in the first place. So perhaps either way you need to read more.

My thinking is that `"alt+c" mode="code"` would not be valid in Proposal 3. The developer would instead have to specify `"alt+KeyC" mode="code"`because `KeyC` is a valid `code` value and `c` is not. Moreover, because the set of valid `key` values and valid `code` values are non-overlapping, having an explicit mode as in Proposal 3 is redundant since it can be inferred from the provided key combo string.

**Comment by bendemboski** [comment]
I think it's definitely good to consider this question, but I'm still finding myself in favor of not having the mode argument.

Let's think about how users are going to first encounter this. If they read the documentation first because they are trying to start using the addon, then no problem - the documentation can be made clear.

If they encounter a usage of it in the wild, there are two possible cases:

1. Wherever they encounter it uses one or the other form consistently
2. Wherever they encounter it uses a mix of the two forms

In case 1, having an extra `mode=` makes it less clear and obvious - I start wondering what mode means only to find that it doesn't change my understanding of what's going on - my initial thought that it's just handing a c keypress was correct and it hasn't aided my understanding or learning other than to force me to learn something that I don't need to know right now. Moreover, there's the risk that I don't go and learn what mode means, and then copy and paste a usage with `mode="code"` (ignoring it because I don't understand or notice it) and change 'KeyUp' to 'w' and suddenly it doesn't work because I have the wrong mode.

In case 2 having `mode=` is potentially helpful because it makes it really clear that there are two modes... But even here the value is limited. As soon as I see the two different forms without the extra mode argument, if I'm confused I'll check the documentation and learn what I need to know, and if I'm not confused then I don't have an extra question to answer.

I think case 1 is going to be the much more common case, so given that the advantages `mode=` provides for case two are pretty minor, and it has significant drawbacks for our probably more common case 1, I don't think it's the way to go.

**Comment by lukemelia** [comment]
@bendemboski Great breakdown and scenario gaming. 👍

**Comment by lukemelia** [comment]
Last call for comments. I think we arrived at a great plan and I intend to start implementing it soon and see what happens when the plan collides with reality!

**Comment by lukemelia** [comment]
Implementation has begun. I've added a checklist to the description of this PR. If anyone wants to contribute, or to pair on any of this, let me know!

**Comment by lukemelia** [comment]
I'm just about ready to cut a beta. Now would be an awesome time for a review if anyone feels up to the task.

**Comment by optikalefx** [comment]
I'm game, I'll put some time on the calendar this week.

**Comment by lukemelia** [comment]
@optikalefx Terrific, thanks. I upgraded one of our apps to use this branch and I'm pretty psyched about how the API feels. I'm aiming for Friday to cut 6.0.0.beta.1. Any changes or feedback before (or after) then would be terrific.

**Comment by optikalefx** [comment]
I am unable to get the component usage to work on my side.

```js
import Component from '@ember/component';
import { onKey } from 'ember-keyboard';
import { inject as service } from '@ember/service';
import { action } from '@ember/object';

export default Component.extend({
	router: service(),

	showing: false,
	route: 'app',

	keystrokeHandler: onKey('alt+KeyC', function() {
		const currentPath = this.get('router._router.currentPath');
		const canGo = currentPath.split('.')[0] === this.route;
		const isAdmin = this.get('session.user.isAdmin') || this.route === 'admin';
		if (isAdmin && canGo) {
			this.set('showing', true);
			this.set('target', this.router.currentURL);
		}
	}),

	cancelHandler: action(function() {
		this.set('showing', false);
		this.set('target', null);
	})
});
```

Error is

```
client-switcher-modal.js:13 Uncaught (in promise) TypeError: (0 , _emberKeyboard.onKey) is not a function
```

I've verified in node_modules your new code is there.  On Ember 3.12

I'll poke around a bit more to see if I can pin-point the issue.

**Comment by lukemelia** [comment]
@optikalefx the component usage API hasn't changed yet, so you still need the mixins in order for it to work (or to manually do the things the mixins do). Your example is missing them.

**Comment by optikalefx** [comment]
Same after adding the mixins back in unfortunately.

`export default Component.extend(EKMixin, EKOnInsertMixin, {`

`Uncaught (in promise) TypeError: (0 , _emberKeyboard.onKey) is not a function`

How does the new modifier API work? Is that supposed to be exported from the addon? I didn't find that in the index file.

**Comment by lukemelia** [comment]
@optikalefx You can find docs that are current to this branch here: https://github.com/adopted-ember-addons/ember-keyboard/blob/rfc/api-change/tests/dummy/app/templates/usage.hbs

The new Javascript API contemplated in the API design doc has not yet been implemented, though it will use the new matching logic under the hood. All the template APIs in the API design docs should work.

I started to look into the Javascript side, but I'm feeling pretty challenged about how we can implement it for Glimmer components, which don't support mixins. Class decorators might provide a path, but I'm not sure if that is a stable API at this point.

**Comment by bendemboski** [comment]
@lukemelia I guess I wasn't thinking about this when reviewing the API design, but I think trying to have a Javascript-only API isn't the way to go. Glimmer components and Octane in general are moving away from doing DOM integration in Javascript, as evidenced primarily by the fact that without *some* support from the template, the Javascript can't access any DOM at all other than via the `document` global, not to mention the lack of a built-in `didInsertElement` equivalent.

I would very much recommend moving away from that, at least for glimmer components. There's just nothing out there in glimmer components (that I know of) that lets you do DOM stuff by just putting some decorators in the Javascript, so even if we could find some way to make it work, I think it would end up being a very unusual and unfamiliar API as Glimmer components/Octane become the norm.

What are the intended benefits of a Javascript-only API in the world of glimmer components?

**Comment by optikalefx** [comment]
So maybe it's useful for me to talk about the use case I tested. I'm making application-side shortcuts. In that way, I can't trust the current element that has focus to capture the keyboard input.

So I guess for this use-case it's about window level key events. I'm not sure I would do that with handlebars only. Perhaps on the top level component? I can test that out.

**Comment by bendemboski** [comment]
@optikalefx that's the use case I was describing in [this comment](https://github.com/adopted-ember-addons/ember-keyboard/pull/135#issuecomment-627475607), and my main motivation for arguing to split out the key event parsing/recognizing from the DOM event attaching. We already have `ember-on-helper` that makes it really easy to attaching document- or window-level event handlers in the context of a component, and it sounds like that's exactly what you're looking for, right?

**Comment by optikalefx** [comment]
Yea, that helper should make things easier. I can test that out here to see if that makes this work from hbs without the JS side.

Maybe ember is going this way, but I would not expect a document level event to live in hbs. That seems like a strange place, and a bit of mixing too much into the templates.

**Comment by bendemboski** [comment]
*shrug* maybe different ways of seeing things. I think it's a really clean story to say that DOM-related stuff goes in templates, and if you want to do it in Javascript, you generally need to call an action from the template. We've seen that with all the magical `click()` and `keyPress()` component methods going away in favor of the `{{on}}` modifier, and the need to use modifiers to access DOM elements from Javascript rather than `this.element.querySelector()`.

In my mind it's much cleaner to know that all DOM event handling that happens in your Javascript is traceable back to the template. It's a nice declarative way of saying "here are the DOM events that I handle" all in one file. Then anytime you see something handling DOM events, you just look in the template to see where it's called, and it will tell you if it's called from an event handler attached to this `<input>`, or that `<button>`, or the document or the window. The alternative is seeing code handling a DOM event, and wondering if it's called from the template, or if I have to search through Javascript for code manually attaching a DOM event listener.

**Comment by optikalefx** [comment]
I see what you're saying. But I think the difference is that in the case of click() that's on a DOM element, so `{{on` makes semantic sense. But with `{{on document}}` that's just random. It's not an element. It's like React putting their router as HTML tags.

**Comment by lukemelia** [comment]
@optikalefx @bendemboski Good discussion, and probably a topic that would be helpful to clarify in our docs.

On the whole, I think @bendemboski captures the Octane-era perspective accurately. One key thing to note in response to one of @optikalefx's concerns is that ember-keyboard *always* uses a document level listener. I do think that in an Octane app using ember-keyboard 6+, I would expect to find global app shortcuts expressed declaratively in application.hbs using the `on-key` helper. e.g.

```hbs
// templates/application.hbs

{{on-key 'ctrl+shift+1' (route-action 'goToMessages')}}
{{on-key 'ctrl+shift+2' (route-action 'goToBrowse')}}
// etc
```

I've been super-happy refactoring my keyboard shortcut usage from the mixin style to `on-key` helpers and modifiers, and I expect that will be how most refactors will go for upgrading ember-keyboard users.

That said, I would like ember-keyboard to expose its functionality via a Javascript API as well, to allow developers flexibility to wield the library as they see fit. I'm just not sure yet of how to craft a succinct API, particularly in the absence of mixins.

Based on that, my current thought is to ship the next version 6 beta with the new template APIs and with the existing mixins API but using the revised key combo matching engine. If we can come up with an improved, cohesive Javascript API, we could include in 6.1. Future developments in the ember ecosytem could help, including the [invokeHelper RFC](https://github.com/emberjs/rfcs/pull/626) and developments around class decorators.

**Comment by bendemboski** [comment]
Yeah, I guess maybe we're approaching this from different perspectives. I'm not interested in a centralized service to manage key events -- it doesn't really match my use cases, or application architecture preferences. DOM event semantics (bubbling, event target inspection, stop propagation, etc.) allow me to coordinate the various pieces of my application that handle key events in a familiar (outside Ember) and easy-to-manage way in all of the use cases I've encountered.

So really what I'm interested in from `ember-keyboard` is the core logic for parsing DOM events and matching them against friendly-readable key combo specifiers. I can plug that into existing modifiers/helpers in a way that seems very familiar and intuitive (from an Ember/Octane perspective).

I'm curious, what are the benefits that y'all get from having a centralized keyboard service?

**Comment by lukemelia** [comment]
>So really what I'm interested in from ember-keyboard is the core logic for parsing DOM events and matching them against friendly-readable key combo specifiers. I can plug that into existing modifiers/helpers in a way that seems very familiar and intuitive (from an Ember/Octane perspective).

I hope that the `if-key` implementation addresses that use case for you.

Re the centralized keyboard service benefits, two thoughts:

1) we have a modal layer in our app and when it is open, I want only keyboard shortcuts for that layer to take precedence. ember-keyboard's `priority` features make this pretty elegant.
2) keyboard events are fundamentally different from clicks/touches/focus to me in that they are only sometimes related to whatever DOM element happens to be focused on the moment. Handling them at a document level and dispatching them accordingly makes more sense to me in consideration of this.

**Comment by bendemboski** [comment]
> I hope that the if-key implementation addresses that use case for you.

Yeah, totally, that's perfect for my use case. I haven't had a chance to try out the new `ember-keyboard`, but I'm very much suspecting it will hit my use cases squarely on the head.

> we have a modal layer in our app and when it is open, I want only keyboard shortcuts for that layer to take precedence.

Yeah, that's good point. I have a modal layer also, but don't have any overlap in the keys it handles and the keys the rest of the application handles, so I haven't had to solve this. A centralized dispatcher with priorities is definitely a good solution, although due to my personal architectural preferences, I'd rather express the dispatching in some explicit logic somewhere, rather than have it be the effect of a bunch of different priority values.

> keyboard events are fundamentally different from clicks/touches/focus to me in that they are only sometimes related to whatever DOM element happens to be focused on the moment

Right, okay, this is kind of a generalization of (1), right? Because a centralized service isn't needed to handle events at the document level -- `{{on-document}}` does that -- just to do the dispatching when multiple consumers want to handle the same event...right? So it seems that is the true value that the service delivers, of resolving multiple consumers that handle the same key combo?

**Comment by optikalefx** [comment]
I think the reason I gravitated towards EK in the first place was that I could add a global event that would delegated to the component I'm in. 

I have an application wide short cut that shows a component - I don't want to define that in `application.hbs`. I want that defined inside that component. You pull up that component file, and right there you see the registered keyboard shortcut to open it.

I'm standing by my feelings that `{{on-document}}` doesn't make sense, and is really headed towards react-land-everything-in-html. But agree to disagree on the overall direction of Octane.

Hopefully I'll get some time tomorrow to test some of the template tools in the beta. Love this discussion, thanks for everyone's input. It's actually been very helpful to show my team at work that there is value to take what we learn and talk it about it in the community.

**Comment by lukemelia** [comment]
> Right, okay, this is kind of a generalization of (1), right? Because a centralized service isn't needed to handle events at the document level -- {{on-document}} does that -- just to do the dispatching when multiple consumers want to handle the same event...right? So it seems that is the true value that the service delivers, of resolving multiple consumers that handle the same key combo?

Hmm, that may be true. I'll have to give that some thought. Perhaps there is some inertia here of using old-school Ember Views/Components as lifecycle management that is no longer necessary with the template helpers.

It also strikes me that we could extract an `ember-key-combo` library that contains just the pieces you are interested in and consume it from `ember-keyboard`.

**Comment by lukemelia** [comment]
> I have an application wide short cut that shows a component - I don't want to define that in application.hbs. I want that defined inside that component. You pull up that component file, and right there you see the registered keyboard shortcut to open it.

How does the code to register the shortcut get executed if the component is only shown once the shortcut is pressed? Is it always rendered and just made visible by the shortcut?

**Comment by optikalefx** [comment]
> How does the code to register the shortcut get executed if the component is only shown once the shortcut is pressed? Is it always rendered and just made visible by the shortcut?

Yes exactly. We have a `<MyComponent>` that wraps our `<AppDialog>`. And in `<MyComponent>` we have the shortcuts defined, which toggle `isShowing` property of the dialog inside.

**Comment by NullVoxPopuli** [comment]
Re Js side of things, maybe use/resource would be best at replacing mixins?

```ts
@use _copy = useKeyShortcut('Ctrl+C'');
// I don't think we need to return anything, but I think @use is required atm. Probs something to mention on the rfc
```

Example implementation here: https://github.com/LevelbossMike/ember-statecharts/pull/241

**Comment by bendemboski** [comment]
> Hmm, that may be true. I'll have to give that some thought. Perhaps there is some inertia here of using old-school Ember Views/Components as lifecycle management that is no longer necessary with the template helpers.

Perhaps. As you illustrated, there's clearly value in centralized logic to dispatch key events when multiple consumers handle the same key combo, because the DOM can't do that for you in all cases. I personally don't think there's any value (other than allowing for inertia) in providing a Javascript-only API to that centralized logic, except extreme cases that I suspect are all anti-patterns where, like, an addon wants to handle keys when it doesn't necessarily render anything but somehow known enough to integrate into the dispatcher's priority mechanism.

> It also strikes me that we could extract an ember-key-combo library that contains just the pieces you are interested in and consume it from ember-keyboard.

Yeah, although I'd probably go a step further and make it a pure Javascript library that implements the DOM-event-to-key-combo-string recognition. What I need on top of that to integrate into Ember is so extremely minimal, that if it were me I'd probably opt to encode that logic in an Ember-independent way (because it really has nothing to do with Ember). I suppose there could be a middle-level addon that just exposes the `{{on-key}}` helper and then `ember-keyboard` with the whole service, but that seems a little excessive 🤷

**Comment by bendemboski** [comment]
> Yes exactly. We have a `<MyComponent>` that wraps our `<AppDialog>`. And in `<MyComponent>` we have the shortcuts defined, which toggle isShowing property of the dialog inside.

FWIW, I'd probably implement it like this:

```hbs
{{#if this.isShowing}}
  {{! component markup }}
{{else}}
  {{on-document (if-key 'ctrl-s' (fn (mut this.isShowing) true)}}
{{/let}}
```

Then the component might not even need a Javascript class.

**Comment by lukemelia** [comment]
@optikalefx I have now implemented the Javascript API from the API design doc, with a few changes. Try it out and let me know how it works for you.

**Comment by lukemelia** [comment]
@NullVoxPopuli wrote:
>Re Js side of things, maybe use/resource would be best at replacing mixins?

I don't think the current architecture of ember-keyboard lends itself well to the use/resource approach. Could be something keep an eye on for the future though.

LMK what you think of the approach I took to implementing the decorator API. These are my first decorators so my confidence is shaky.

File: addon/decorators/key-responder.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+export default function keyResponder(DecoratedClass, opts = {}) {
+  if (opts.priority === undefined) {
+    opts.priority = 0;
+  }
+
+  if (opts.activated === undefined) {
+    opts.activated = true;
+  }
+
+  return class ClassAsKeyResponder extends DecoratedClass {
```
Comment (NullVoxPopuli) [inline addon/decorators/key-responder.js:30]: this should probably have a `static name = `${DecoratedClass.name}WithKeyResponder`;

Comment (lukemelia) [reply addon/decorators/key-responder.js:30]: Thanks!

**Comment by lukemelia** [comment]
OK, I think we are good to go for a new beta. I'll leave this for a day or two before I merge in case anyone wants to do some more code review.

**Comment by NullVoxPopuli** [comment]
Just upgraded emberclear to v6. Nice work everyone! :D
upgrading was easy, and I was able to delete a bunch of my own code around handling key-events! :tada:

**Comment by lukemelia** [comment]
@NullVoxPopuli Great to hear! 👍

**Pull request merged**